File: python/mongo-app.py
#
# Copyright IBM Corp. 2014
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
from flask import Flask, jsonify, abort, request
from pymongo import MongoClient
from bson import Binary, Code, ObjectId
from bson.json_util import dumps
from apscheduler.scheduler import Scheduler
import os
import json

app = Flask(__name__, static_url_path='/static')

# Gets the Mongo DB URL and DB from VCAP_SERVICES if present, else
# assumes the Mongo instance is running locally
url = 'mongodb://localhost:27017/'
dbName = 'db'
if os.environ.has_key('VCAP_SERVICES'):
    vcapJson = json.loads(os.environ['VCAP_SERVICES'])
    for key, value in vcapJson.iteritems():
        #Only find the services with the name todo-mongo-db, there should only be one
        mongoServices = filter(lambda s: s['name'] == 'todo-mongo-db', value)
        if len(mongoServices) != 0:
            mongoService = mongoServices[0]
            url = mongoService['credentials']['url']
            dbName = mongoService['credentials']['db']

client = MongoClient(url)
db = client[dbName]
collection = db.todos

# Create a scheduled task that will cleanup ToDos if the list grows too big
sched = Scheduler()
sched.start()
def cleanup():
    if collection.count() > 20:
        collection.remove(collection.find_one())
sched.add_interval_job(cleanup, seconds=30)

# Serve up index.html when requests come to root
@app.route('/')
def index():
    return app.send_static_file('index.html')

# Serve up static resources, JavaScript, HTML, CSS
@app.route('/<path:path>')
def js_proxy(path):
    # send_static_file will guess the correct MIME type
    return app.send_static_file(os.path.join(path))

#Helper function to preprocess Mongo ToDos to what the client expects
def to_todo_obj(todo):
    todo['id'] = str(todo['_id'])
    del todo['_id']
    return todo

@app.route('/api/todos', methods = ['GET'])
def get_todo():
	return dumps(map(to_todo_obj, collection.find()))

@app.route('/api/todos', methods = ['POST'])
def create_todo():
    if not request.json:
        abort(400)
    todo = request.json
    id = collection.insert(todo)
    todo['id'] = str(id)
    del todo['_id']
    return dumps(todo)

@app.route('/api/todos/<id>', methods = ['DELETE'])
def delete_todo(id):
    collection.remove({'_id' : ObjectId(id)})
    return '', 204

@app.route('/api/todos/<id>', methods = ['PUT'])
def update_todo(id):
    if not request.json:
        abort(400)
    todo = collection.find_one({'_id' : ObjectId(id)})
    newToDo = request.json
    todo['title'] = newToDo['title']
    todo['order'] = newToDo['order']
    todo['completed'] = newToDo['completed']
    collection.save(todo)
    del todo['_id']
    return dumps(todo)

port = os.getenv('VCAP_APP_PORT', '5000')
if __name__ == '__main__':
    app.run(host='0.0.0.0', port=int(port), debug = True)

File: php/MongoApp.php
<?php
/*
 * Copyright IBM Corp. 2014
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 /**
  * Class to handle performing basic CRUD operations to a Mongo
  * DB.  This class assumes you have the Mongo DB driver extension
  * installed in PHP.
  */
final class MongoApp {
	private static $inst = null;
	private $mongoUrl = 'mongodb://localhost:27017';
	private $dbName = 'db';

    public static function Instance() {
        if (self::$inst === null) {
            self::$inst = new MongoApp();
        }
        return self::$inst;
    }

    private function __construct() {
		if($vcapStr = getenv('VCAP_SERVICES')) {
			$vcap = json_decode($vcapStr, true);
			foreach ($vcap as $serviceTypes) {
				foreach ($serviceTypes as $service) {
					if($service['name'] == 'todo-mongo-db') {
						$credentials = $service['credentials'];
						$this->mongoUrl = $credentials['uri'];
						$parsedUrl = parse_url($credentials['uri']);
						$this->dbName = preg_replace('/\/(.*)/', '$1', $parsedUrl['path']);
						break;
					}
				}
			}
		}
    }

    private function getConnection() {
		return new Mongo($this->mongoUrl);
	}

	private function getTodosCollection($conn) {
		$name = $this->dbName;
		$mongodb = $conn->$name;
    	return $mongodb->todos;
	}

	/**
	 * Transforms the ToDo JSON from the Mongo DB to the JSON
	 * the client will expect.
	 */
	private function toClientToDo($mongoTodo) {
		$mongoTodo['id'] = $mongoTodo['_id']->{'$id'};
		unset($mongoTodo['_id']);
		return $mongoTodo;
	}

	/**
	 * Gets all ToDos from the DB.
	 */
	public function get() {
		$conn = $this->getConnection();
    	$collection = $this->getTodosCollection($conn);
		$cursor = $collection->find();
		$result = array();
		foreach ($cursor as $todo) {
			$result[] = $this->toClientToDo($todo);
		}
		$conn->close();
		return $result;
	}

	/**
	 * Creates a new ToDo in the DB.
	 */
	public function post($todo) {
		$conn = $this->getConnection();
    	$collection = $this->getTodosCollection($conn);
    	$collection->save($todo);
    	$conn->close();
		return $this->toClientToDo($todo);
	}

	/**
	 * Updates a ToDo in the DB.
	 */
	public function put($id, $todo) {
		$conn = $this->getConnection();
    	$collection = $this->getTodosCollection($conn);
    	$mongoTodo = $collection->findOne(array('_id' => new MongoId($id)));
    	$mongoTodo['title'] = $todo['title'];
    	$mongoTodo['completed'] = $todo['completed'];
    	$mongoTodo['order'] = $todo['order'];
    	$collection->save($mongoTodo);
    	$conn->close();
    	return $this->toClientToDo($mongoTodo);
	}

	/**
	 * Deletes a ToDo from the DB.
	 */
	public function delete($id) {
		$conn = $this->getConnection();
    	$collection = $this->getTodosCollection($conn);
    	$collection->remove(array('_id' => new MongoId($id)));
    	$conn->close();
	}
}